src/batch/src/executor/generic_exchange.rs

```diff
@@ -165,27 +168,53 @@ impl<C: BatchTaskContext> GenericExchangeExecutor<C> {
 }
 
 #[try_stream(boxed, ok = DataChunk, error = RwError)]
-async fn data_chunk_stream(mut source: ExchangeSourceImpl, metrics: Option<BatchTaskMetrics>) {
+async fn data_chunk_stream<C: BatchTaskContext>(
+    mut source: ExchangeSourceImpl,
+    context: C,
+    identity: String,
+) {
+    // create the collector
+    let source_id = source.get_task_id();
+    let counter = {
+        let mut labels = context.task_labels();
+        labels.insert("executor_id".to_string(), identity.clone());
+        labels.insert(
+            "source_stage_id".to_string(),
+            source_id.stage_id.to_string(),
+        );
+        labels.insert("source_task_id".to_string(), source_id.task_id.to_string());
+
+        let opts = Opts::new(
+            "batch_exchange_recv_row_number",
+            "Total number of row that have been received from upstream source",
+        )
+        .const_labels(labels);
+        let counter = IntCounter::with_opts(opts).unwrap();
+        match context.register(Box::new(counter.clone())) {
+            Ok(_) => Some(counter),
+            Err(_) => None,
```

Comment:
We should return err here rather `None`, this is a logic error.

Reply:
You mean in this way ?
```
let counter:Result<Counter, BatchError> = {
 ...
  Ok(_) => Ok(counter),
  Err(e) => Err(e)
};
```

Reply:
We can throw this error out using `?`.

Reply:
> It returns an error if the descriptors provided by the collector are invalid or if they — in combination with descriptors of already registered Collectors — do not fulfill the consistency and uniqueness criteria described in the documentation of desc.

If we throw it out, the execution will be stop. I think we can't stop the execution because of the failure of creating the metrics collector?